exercises.ipynb

```diff
@@ -2,125 +2,1002 @@
+    "\n",
+    "#get final page\n",
+    "last_url = soup.select('li.last a')[0]['href']\n",
+    "\n",
+    "#create list of all pages to parse\n",
+    "url_list = [last_url[:-2]+str(page) for page in range(1,int(last_url[-2:])+1)]"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 65,
+   "metadata": {},
    "outputs": [],
    "source": [
     "def get_titles(soup):    \n",
     "    \"\"\" Returns a list of titles on the page \"\"\"\n",
-    "    # the \"soup\" parameter is of the type that is\n",
-    "    # returned by Beautiful Soup when it parses HTML.\n",
-    "    # The function should then use the object to\n",
-    "    # extract a list of titles (of the lego sets)\n",
-    "    #\n",
-    "    # Lookup the documentation for Beautiful Soup\n",
-    "    # Figure out how to select the text of the title\n",
-    "    # of each legoset. A title should look like: \n",
-    "    # \"10252: Volkswagen Beetle\"\n",
-    "    pass\n",
+    "\n",
+    "    #Use Beautiful Soup select with list comprehension\n",
+    "    titles = [soup.select('div.meta h1 a')[i].text for i in range(len(soup.select('div.meta h1 a')))]\n",
+    "\n",
+    "    return titles\n",
     "\n",
     "def parse_bricks(url):\n",
     "    \"\"\" Fetches Lego Bricks page and extracts titles \"\"\"\n",
-    "    # Lookup the documentation to the \"requests\" library\n",
-    "    #\n",
-    "    # Use requests to make a get request to the\n",
-    "    # url given in the argument \"url\" (which is a string)\n",
-    "    # and get the raw HTML body of the response\n",
-    "    #\n",
-    "    # Use \"BeautifulSoup\" to parse this HTML. \n",
+    "\n",
+    "    #get url request\n",
+    "    req = requests.get(url)\n",
+    "    \n",
+    "    # Use \"BeautifulSoup\" to parse this HTML.\n",
+    "    \n",
+    "    soup = BeautifulSoup(req.text, \"html.parser\")\n",
     "    #\n",
     "    # Use the \"get_titles\" function to extract the\n",
     "    # titles from the BeautifulSoup object.\n",
+    "    \n",
+    "    titles = get_titles(soup)\n",
     "    #\n",
     "    # Return the titles\n",
-    "    pass"
+    "    return titles"
    ]
   },
   {
    "cell_type": "code",
-   "execution_count": 276,
-   "metadata": {
-    "autoscroll": false,
-    "collapsed": false,
-    "ein.hycell": false,
-    "ein.tags": "worksheet-0",
-    "slideshow": {
-     "slide_type": "-"
+   "execution_count": 66,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "CPU times: user 6.73 s, sys: 34.2 ms, total: 6.77 s\n",
+      "Wall time: 50.2 s\n"
+     ]
     }
-   },
+   ],
+   "source": [
+    "%%time\n",
+    "#go page by page to get titles - combine back into one list\n",
+    "bricks = []\n",
+    "for url in url_list:\n",
+    "    bricks.append(parse_bricks(url))\n",
+    "    \n",
+    "bricks = list(itertools.chain.from_iterable(bricks))"
```

Comment:
I think you're just flattening a list? Not bad, but no need for itertools, very canonical to just use for comprehension: 

```
[i for x in y for i in x]
```

Reply:
(in fact, you don't need the for loop, it's a pure operation (no side affects), so canonical to use for comprehension itself!